File: ios/Sources/Scoring.swift
// Scoring für den Klassik-Modus (relative Ordnung) — portiert aus src/lib/scoring.ts.
// Gleiche Jahre an den Rändern gelten als korrekt (>=/<=), Ties sind also erlaubt.
import Foundation

enum Scoring {
    /// Karten chronologisch nach Jahr sortieren (Swift-Sort ist stabil).
    static func sortByYear(_ cards: [Track]) -> [Track] {
        cards.sorted { $0.releaseYear < $1.releaseYear }
    }

    /// insertIndex ist die Lücke: 0 = ganz links, cards.count = ganz rechts.
    static func isPlacementCorrect(sorted: [Track], track: Track, insertIndex: Int) -> Bool {
        let left = insertIndex > 0 ? sorted[insertIndex - 1] : nil
        let right = insertIndex < sorted.count ? sorted[insertIndex] : nil
        let okLeft = left.map { track.releaseYear >= $0.releaseYear } ?? true
        let okRight = right.map { track.releaseYear <= $0.releaseYear } ?? true
        return okLeft && okRight
    }

    /// Korrekt platzierte Karte einsortiert zurückgeben (bleibt chronologisch).
    static func insertCard(sorted: [Track], track: Track) -> [Track] {
        sortByYear(sorted + [track])
    }

    struct PlayerStats: Identifiable {
        var id: UUID { playerId }
        let playerId: UUID
        let name: String
        let colorHex: String
        let cards: Int
        let accuracy: Double // 0..1
    }

    static func ranking(_ players: [Player]) -> [PlayerStats] {
        players
            .map {
                PlayerStats(
                    playerId: $0.id,
                    name: $0.name,
                    colorHex: $0.colorHex,
                    cards: $0.cards.count,
                    accuracy: $0.attempts > 0 ? Double($0.hits) / Double($0.attempts) : 0
                )
            }
            .sorted { ($0.cards, $0.accuracy) > ($1.cards, $1.accuracy) }
    }
}

File: src/lib/scoring.ts
// Scoring für den Klassik-Modus (relative Ordnung).
// Siehe docs/game-design.md → Grundregeln + Edge-Case „Zwei Karten aus dem gleichen Jahr".
import type { Player, Track } from '../types';

/** Karten chronologisch nach Jahr sortieren (stabil). */
export function sortByYear(cards: Track[]): Track[] {
  return [...cards].sort((a, b) => a.releaseYear - b.releaseYear);
}

/**
 * Ist die Platzierung in der (bereits sortierten) Timeline korrekt?
 * insertIndex ist die Lücke: 0 = ganz links, cards.length = ganz rechts.
 * Gleiche Jahre an den Rändern gelten als korrekt (>=/<=), Ties sind also erlaubt.
 */
export function isPlacementCorrect(
  sortedCards: Track[],
  track: Track,
  insertIndex: number,
): boolean {
  const left = insertIndex > 0 ? sortedCards[insertIndex - 1] : null;
  const right = insertIndex < sortedCards.length ? sortedCards[insertIndex] : null;
  const okLeft = left === null || track.releaseYear >= left.releaseYear;
  const okRight = right === null || track.releaseYear <= right.releaseYear;
  return okLeft && okRight;
}

/** Korrekt platzierte Karte einsortiert zurückgeben (bleibt chronologisch). */
export function insertCard(sortedCards: Track[], track: Track): Track[] {
  return sortByYear([...sortedCards, track]);
}

/** Optionaler Jahres-Modus: Schätzung innerhalb ±tolerance Jahre. */
export function isYearGuessCorrect(
  track: Track,
  guessedYear: number,
  tolerance: number,
): boolean {
  return Math.abs(track.releaseYear - guessedYear) <= tolerance;
}

export type PlayerStats = {
  playerId: string;
  name: string;
  cards: number;
  accuracy: number; // 0..1
  bonusPoints: number; // "Wessen Liebling?" / "Artist & Titel raten"
};

export function ranking(players: Player[]): PlayerStats[] {
  return players
    .map((p) => ({
      playerId: p.id,
      name: p.name,
      cards: p.cards.length,
      accuracy: p.attempts > 0 ? p.hits / p.attempts : 0,
      bonusPoints: p.bonusPoints ?? 0,
    }))
    // Kartenzahl bleibt die Sieg-Metrik; Bonuspunkte zählen nur als Tiebreaker.
    .sort((a, b) => b.cards - a.cards || b.accuracy - a.accuracy || b.bonusPoints - a.bonusPoints);
}
